sdk/cpp/core/src/gnmi_provider.cpp

```diff
@@ -0,0 +1,293 @@
+#include "logger.hpp"
+
+using grpc::Channel;
+
+using namespace std;
+using namespace ydk;
+
+namespace ydk
+{
+	//const char* CANDIDATE = "urn:ietf:params:netconf:capability:candidate:1.0";
+	static path::SchemaNode* get_schema_for_operation(path::RootSchemaNode& root_schema, string operation);
+
+	static shared_ptr<path::Rpc> create_rpc_instance(path::RootSchemaNode & root_schema, string rpc_name);
+	static path::DataNode& create_rpc_input(path::Rpc & gnmi_rpc);
+	static void create_input_source(path::DataNode & input, bool config);
+
+	static string get_read_rpc_name(bool config);
+	static bool is_config(path::Rpc & rpc);
+	static string get_filter_payload(path::Rpc & ydk_rpc);
+	static string get_gnmi_payload(path::DataNode & input, string data_tag, string data_value);
+	static std::shared_ptr<path::DataNode> handle_read_reply(string reply, path::RootSchemaNode & root_schema);
+
+	gNMIServiceProvider::gNMIServiceProvider(string address, string username, string password)
+		: client(make_unique<gNMIClient>(grpc::CreateChannel(address, grpc::InsecureChannelCredentials())))
+	{
+		path::Repository repo;
+		initialize(repo);
+		YLOG_INFO("Connected to {} on port {} using ssh", address);
+	}
+
+	gNMIServiceProvider::gNMIServiceProvider(path::Repository & repo, string address, string username, string password)
+		: client(make_unique<gNMIClient>(grpc::CreateChannel(address, grpc::InsecureChannelCredentials())))
+	{
+		initialize(repo);
+		YLOG_INFO("Connected to {} on port {} using ssh", address);
+	}
+
+	gNMIServiceProvider::~gNMIServiceProvider() = default;
+
+	void gNMIServiceProvider::initialize(path::Repository & repo) {
+		IetfCapabilitiesParser capabilities_parser{};
+		client->connect();
+		server_capabilities = client->Capabilities();
+		
+		root_schema = repo.create_root_schema(capabilities_parser.parse(server_capabilities));
+
+		if(root_schema.get() == nullptr)
+		{
+			YLOG_ERROR("Root schema cannot be obtained");
+			throw(YCPPIllegalStateError{"Root schema cannot be obtained"});
+		}
+	}
+
+	EncodingFormat gNMIServiceProvider::get_encoding() const
+	{
+		return EncodingFormat::JSON;
+	}
+
+	path::RootSchemaNode& gNMIServiceProvider::get_root_schema() const
+	{
+		return *root_schema;
+	}
+
+	std::shared_ptr<path::DataNode> gNMIServiceProvider::invoke(path::Rpc& rpc) const
+	{
+		path::SchemaNode* read_schema = get_schema_for_operation(*root_schema, "ydk:read");
+
+		//for now we only support crud rpc's
+    	path::SchemaNode* rpc_schema = &(rpc.schema());
+		//print_paths(*rpc_schema);
+
+		std::shared_ptr<path::DataNode> datanode = nullptr;
+		
+		if(rpc_schema == read_schema)
+	    {
+	    	//print_data_paths(rpc.input());
+	        return handle_read(rpc);
+	    }
+	    else
+	    {
+	        YLOG_ERROR("rpc is not supported");
+	        throw(YCPPOperationNotSupportedError{"rpc is not supported!"});
+	    }
+
+	    return datanode;
+	}
+
+	void gNMIServiceProvider::print_paths(ydk::path::SchemaNode& sn) const
```

Comment:
This looks like it could be meant for `debugging` purpose, not sure if it should go into the code?